pubsub/rocketmq/rocketmq.go

```diff
@@ -16,120 +16,247 @@ package rocketmq
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
+	"strconv"
+	"strings"
 	"sync"
 	"time"
 
 	mq "github.com/apache/rocketmq-client-go/v2"
 	mqc "github.com/apache/rocketmq-client-go/v2/consumer"
 	"github.com/apache/rocketmq-client-go/v2/primitive"
 	mqp "github.com/apache/rocketmq-client-go/v2/producer"
-	"github.com/cenkalti/backoff/v4"
+	"github.com/apache/rocketmq-client-go/v2/rlog"
 
+	"github.com/dapr/components-contrib/internal/utils"
 	"github.com/dapr/components-contrib/pubsub"
 	"github.com/dapr/kit/logger"
-	"github.com/dapr/kit/retry"
 )
 
-type topicData struct {
-	selector      mqc.MessageSelector
-	handler       pubsub.Handler
-	consumerGroup string
-	mqType        string
-	mqExpr        string
+type daprQueueSelector struct {
+	hashQueueSelector       mqp.QueueSelector
+	roundRobinQueueSelector mqp.QueueSelector
+}
+
+func NewDaprQueueSelector() *daprQueueSelector {
+	return &daprQueueSelector{
+		hashQueueSelector:       mqp.NewHashQueueSelector(),
+		roundRobinQueueSelector: mqp.NewRoundRobinQueueSelector(),
+	}
+}
+
+func (p *daprQueueSelector) Select(msg *primitive.Message, queues []*primitive.MessageQueue) *primitive.MessageQueue {
+	if msg.Queue != nil {
+		return msg.Queue
+	}
+	if queue := msg.GetProperty(metadataRocketmqQueue); queue != "" {
+		for _, q := range queues {
+			if strconv.Itoa(q.QueueId) == queue {
+				return q
+			}
+		}
+	}
+	key := msg.GetShardingKey()
+	if len(key) == 0 {
+		return p.roundRobinQueueSelector.Select(msg, queues)
+	}
+	return p.hashQueueSelector.Select(msg, queues)
 }
 
 type rocketMQ struct {
-	name     string
-	metadata *rocketMQMetaData
-
-	logger       logger.Logger
-	topics       map[string]topicData
-	producer     mq.Producer
-	producerLock sync.RWMutex
-	consumer     mq.PushConsumer
-	consumerLock sync.RWMutex
-
-	ctx    context.Context
-	cancel context.CancelFunc
+	name          string
+	metadata      *rocketMQMetaData
+	producer      mq.Producer
+	producerLock  sync.RWMutex
```

Comment:
Looks like you're not using RLocks, so can be changed to `sync.Mutex`

Reply:
ok